src/metabase/models/recent_views.clj

```diff
@@ -155,80 +162,110 @@
-  querying the database for the object."
+  For most things, we try to gather the information in a single query, but for cards and datasets, but certain things
+  are way easier to check with code (e.g. a collection's parent collection.)"
   (fn [{:keys [model #_model_id #_timestamp card_type]}]
-    (or (get {"model" :dataset
-              "question" :card} card_type)
+    (or (get {"model" :dataset "question" :card} card_type)
         (keyword model))))
 
 (defmethod fill-recent-view-info :default [m] (throw (ex-info "Unknown model" {:model m})))
 
-(defn get-parent-coll
+(defn- get-parent-coll*
   "Gets parent collection info for a recent view item."
-  [coll-id-or-coll]
-  (-> (cond (map? coll-id-or-coll) (if-let [parent-id (:parent_id (t2/hydrate coll-id-or-coll :parent_id))]
-                                     ;; hydrate the effective location on the collection
-                                     (t2/select-one :model/Collection parent-id)
-                                     (root/root-collection-with-ui-details {}))
-            (nil? coll-id-or-coll) (root/root-collection-with-ui-details {})
-            :else (t2/select-one :model/Collection coll-id-or-coll))
-      (select-keys [:id :name :authority_level])
-      (update :authority_level #(some-> % keyword))))
-
-(mu/defn get-moderated-status
-  "Returns moderated_status for a given model and model-id.
-
-  (Currently only used for cards and models, but ought to be extended to dashboards in the future)"
-  [model :- [:enum :card] model-id] :- [:maybe "verified"]
-  (-> (t2/select-one [:model/ModerationReview :status]
-                     {:where [:and
-                              [:= :moderated_item_id model-id]
-                              [:= :moderated_item_type (name model)]
-                              [:= :most_recent true]]})
-      :status))
-
-(defn ellide-archived
-  "Returns nil if the model is archived, otherwise returns the model.
+  ;; user-id is not used, but we need it to memoize the function correctly
+  [coll-id]
+  (if (nil? coll-id)
+    (root/root-collection-with-ui-details {})
+    (-> (t2/select-one :model/Collection coll-id)
+        (select-keys [:id :name :authority_level])
+        (update :authority_level #(some-> % keyword)))))
+
+(def ^{:private true :arglists '([coll-id])}
+  get-parent-coll
+  (memoize/ttl get-parent-coll* :ttl/threshold 1000))
+
+(defn- ellide-archived
+  "Returns the model when it's not archived.
   We use this to ensure that archived models are not returned in the recent views."
   [model]
-  (cond
-    (false? (:archived model)) model
-
-    ;; The model for this recent view doesn't exist in the database, so we can't check if it's archived, so just
-    ;; filter it out.
-    (not model) nil
-
-    (true? (:archived model)) nil
-
-    (nil? (:archived model))
-    (throw (ex-info "Archived field is nil" {:model model}))))
+  (when (false? (:archived model)) model))
```

Comment:
nit - this feels a little more readable to me using a `when-not`
```suggestion
  (when-not (:archived model) model))
```

Reply:
but if `(:archived model)` is nil, I want that function to return nil